transport/ice/IceSession.cpp

```diff
@@ -472,11 +493,252 @@ bool IceSession::isResponseAuthentic(const void* data, size_t len) const
+        remotePort.toString().c_str(),
+        localEndpoint->getLocalPort().toString().c_str());
+
+    int remoteCandidatePriority = 0;
+    const auto prioAttribute = msg.getAttribute<StunPriority>(StunAttribute::PRIORITY);
+    if (prioAttribute)
+    {
+        remoteCandidatePriority = prioAttribute->value;
+    }
+
+    bool successfullyProcessed = false;
+
+    if (localEndpoint->getTransportType() == TransportType::TCP)
+    {
+        successfullyProcessed =
+            processValidStunTcpRequest(localEndpoint, remotePort, msg, remoteCandidatePriority, now);
+    }
+    else if (localEndpoint->getTransportType() == TransportType::UDP)
+    {
+        successfullyProcessed =
+            processValidStunUdpRequest(localEndpoint, remotePort, msg, remoteCandidatePriority, now);
+    }
+
+    if (successfullyProcessed && _credentials.role == IceRole::CONTROLLED)
+    {
+        const auto useCandidate = msg.getAttribute(StunAttribute::USE_CANDIDATE);
+        auto useCandidatePair = useCandidate ? findCandidatePair(localEndpoint, remotePort) : nullptr;
+        if (useCandidatePair)
+        {
+            if (_nomination != useCandidatePair)
+            {
+                logger::debug("remote nominated %s-%s",
+                    _logId.c_str(),
+                    useCandidatePair->localCandidate.address.toString().c_str(),
+                    useCandidatePair->remoteCandidate.address.toString().c_str());
+                if (useCandidatePair->state == ProbeState::Succeeded)
+                {
+                    if (_nomination)
+                    {
+                        // Aggressive nomination
+                        // Then we will clean the old nominated flag
+                        _nomination->nominated = false;
+                    }
+
+                    useCandidatePair->nominate(now);
+                    _nomination = useCandidatePair;
+                    _preliminaryCandidate = useCandidatePair;
+
+                    if (_state == State::CONNECTED && _eventSink)
+                    {
+                        // aggressive nomination
+                        _eventSink->onIceCandidateChanged(this,
+                            useCandidatePair->localEndpoint.endpoint,
+                            useCandidatePair->remoteCandidate.address);
+                    }
+                }
+            }
+            processTimeout(now);
+        }
+    }
+}
+
 void IceSession::onRequestReceived(IceEndpoint* localEndpoint,
     const transport::SocketAddress& remotePort,
     const StunMessage& msg,
     const uint64_t now)
 {
+    if (_state >= State::FAILED)
+    {
+        // SMB does not have a way to restart ICE. Then we should not respond more to probes after we have a failed
+        // state otherwise peer might think the ICE connection is still good and then does not try any recover
+        // mechanism. So if it is FAILED, peer MUST fail too
+        return;
```

Comment:
As SMB does not have mechanism to report its ICE state, it can't respond to stuns when it is in FAIL state, otherwise clients might not fail and then they don't trigger the recover